doc/index.md

```diff
@@ -331,6 +331,157 @@ to inject things into the public interface of the iterator:
+        {
+            return it == end;
+        }
+        // or
+        template<typename I, typename S, typename SA>
+        constexpr bool empty(I const &it, S const &end, SA const &end_adapt) const 
+        {
+            return 
+              *this.some_value == end_adapt.some_value
+              && it == end;
+        }
+        
+        template<typename I>
+        reference_t<I> read(I const &it)
+        {
+            return *it;
+        }
+
+        template<typename I>
+        void next(I &it)
+        {
+            ++it;
+        }
+
+        // !For BidirectionalIterator only!
+        template<typename I>
+        void prev(I &it)
+        {
+            --it;
+        }
+
+        // !For RandomAccessIterator only!
+        template<typename I>
+        void advance(I &it, difference_type_t<I> n)
+        {
+            it += n;
+        }    
+
+        // !For SizedIterator only!
+        template<typename I>
+        difference_type_t<I> distance_to(I const &this_iter, I const &that_iter)
+        {
+            return that_iter - this_iter;
+        }
+        // or
+        template<typename I>
+        difference_type_t<I> distance_to
+            (I const &this_iter, I const &that_iter, adaptor const &that_adapt)
+        {
+            return that_iter - this_iter;
+        }
+    }
+~~~~~~~
+
+As you can see, some "overrides" have effect only for `begin_adaptor` or `end_adaptor`.
+In order to use full potential of adaptor, you need to have separate adaptors for begin and end:
+
+~~~~~~~{.cpp}
+    struct adaptor : adaptor_base
+    {
+        int n = 0;
+
+        void next(iterator_t<Rng>& it) 
+        {
+            ++n;
+            ++it;
+        }
+    };
+
+    struct sentinel_adaptor : adaptor_base 
+    {
+        int stop_at;        
+        bool empty(const iterator_t<Rng>&, const adaptor& ia, const sentinel_t<Rng>& s) const 
+        {
+            return ia.n == stop_at;
+        }
+    };
+
+    adaptor begin_adaptor() const { return {}; }
+    sentinel_adaptor end_adaptor() const { return {100}; }
+~~~~~~~
+
+Sometimes, you can use the same adaptor for both `begin_adaptor` and `end_adaptor`:
+
+~~~~~~~{.cpp}
+    struct adaptor : adaptor_base
+    {
+        int n = 0;
+        void next(iterator_t<Rng>& it) 
+        {
+            ++n;
+            ++it;
+        }
+
+        // pay attention, we use equal, not empty. empty() will never trigger.
+        template<typename I>
+        bool equal(I const &this_iter, I const &that_iter, adaptor const &that_adapt) const
+        {
+            return *this.n == that_adapt.n;
+        }
+    };
+
+    adaptor begin_adaptor() const { return {}; }
+    adaptor end_adaptor()   const { return {100}; }
+~~~~~~~
+
+Pay attentention that all data, that you'll store in adaptor - will become part of iterator payload.
+
+If you will not "override" `begin_adaptor()` or/and `end_adaptor()` in your view_adaptor - default ones will be used.
```

Comment:
```suggestion
If you will not "override" `begin_adaptor()` or/and `end_adaptor()` in your view_adaptor, default ones will be used.
```

Reply:
Applying, but I disagree that commas are better than dashes.